src/Microsoft.AspNetCore.Http/StreamPipeWriter.cs

```diff
@@ -0,0 +1,307 @@
+                        await _writingStream.WriteAsync(segment.Buffer.ToArray(), 0, segment.Length, token);
+#else
+#error Target frameworks need to be updated.
+#endif
+                        segment.Return();
+                    }
+
+                    _completedSegments.Clear();
+                }
+
+                if (!_currentSegment.IsEmpty)
+                {
+#if NETCOREAPP2_2
+                    await _writingStream.WriteAsync(_currentSegment.Slice(0, _position), token);
+#elif NETSTANDARD2_0
+                    await _writingStream.WriteAsync(_currentSegment.ToArray(), 0, _position, token);
+#else
+#error Target frameworks need to be updated.
+#endif
+                }
+
+                await _writingStream.FlushAsync(token);
+            }
+            catch (OperationCanceledException)
+            {
+                // Catch any cancellation and translate it into setting isCanceled = true
+                return new FlushResult(isCanceled: true, IsCompletedOrThrow());
+            }
+
+            // After writing to the stream, we can return all ArrayPool segments that are completed.
+            _bytesWritten = 0;
+            _position = 0;
+
+            return new FlushResult(isCanceled: false, IsCompletedOrThrow());
+        }
+
+        private void EnsureCapacity(int sizeHint)
+        {
+            // This does the Right Thing. It only subtracts _position from the current segment length if it's non-null.
+            // If _currentSegment is null, it returns 0.
+            var remainingSize = _currentSegment.Length - _position;
+
+            // If the sizeHint is 0, any capacity will do
+            // Otherwise, the buffer must have enough space for the entire size hint, or we need to add a segment.
+            if ((sizeHint == 0 && remainingSize > 0) || (sizeHint > 0 && remainingSize >= sizeHint))
+            {
+                // We have capacity in the current segment
+                return;
+            }
+
+            AddSegment(sizeHint);
+        }
+
+        private void AddSegment(int sizeHint = 0)
+        {
+            if (_currentSegment.Length != 0)
+            {
+                // We're adding a segment to the list
+                if (_completedSegments == null)
+                {
+                    _completedSegments = new List<CompletedBuffer>();
+                }
+
+                // Position might be less than the segment length if there wasn't enough space to satisfy the sizeHint when
+                // GetMemory was called. In that case we'll take the current segment and call it "completed", but need to
+                // ignore any empty space in it.
+                _completedSegments.Add(new CompletedBuffer(_currentSegmentOwner, _position));
+            }
+
+            // Get a new buffer using the minimum segment size, unless the size hint is larger than a single segment.
+            _currentSegmentOwner = _pool?.Rent(Math.Max(_minimumSegmentSize, sizeHint));
```

Comment:
Is `_pool` ever null? `_pool?.` if it can be next line will fail.

As aside it will also devirtualize in .NET Core 3.0 https://github.com/dotnet/corefx/pull/33085 so again you could not set it to `MemoryPool<byte>.Shared` in the constructor when `null` and instead here have:
```csharp
var size = Math.Max(_minimumSegmentSize, sizeHint);
_currentSegmentOwner = _pool?.Rent(size) ?? MemoryPool<byte>.Shared.Rent(size);
```

Reply:
I should update this. My intention was to compare MemoryPool<byte>.Shared vs ArrayPool with this change. _pool right now cannot be null. Let me remove the nullable check.